compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNestedClassChecker.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors.NESTED_CLASS_NOT_ALLOWED
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.*
+
+object FirNestedClassChecker : FirRegularClassChecker() {
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        // Local enums / objects / companion objects are handled with different diagnostic codes.
+        if (declaration.classKind.isSingleton && declaration.isLocal) return
```

Comment:
According to https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/resolve/ModifiersChecker.java#L201, enum class, object, and companion object are bailed out here.

`isSingleton` checks if the class kind is enum entry or object. Here, we accidentally bail out for enum entry as well. You better explicitly check if the class kind is `ENUM_CLASS` or `OBJECT`.

Reply:
enum entry bails out on line 198 in ModifiersChecker.java. So this should be equivalent.

Reply:
Okay, using `isSingleton` seems fine in the sense that it can help us bail out for local object.

But, still, you won't need to check if the class kind is `ENUM_CLASS` and bail out too? Note that enum class and enum entries are different.

Reply:
Ah thanks! I missed that.